pkgs/applications/virtualization/qemu/default.nix

```diff
@@ -5,45 +5,53 @@
-, ncursesSupport ? !nixosTestRunner && !toolsOnly, ncurses
+, guestAgentSupport ? (with stdenv.hostPlatform; isLinux || isNetBSD || isOpenBSD || isSunOS || isWindows) && !minimal
+, numaSupport ? stdenv.isLinux && !stdenv.isAarch32 && !minimal, numactl
+, seccompSupport ? stdenv.isLinux && !minimal, libseccomp
+, alsaSupport ? lib.hasSuffix "linux" stdenv.hostPlatform.system && !nixosTestRunner && !minimal
+, pulseSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, libpulseaudio
+, pipewireSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, pipewire
+, sdlSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, SDL2, SDL2_image
+, jackSupport ? !stdenv.isDarwin && !nixosTestRunner && !minimal, libjack2
+, gtkSupport ? !stdenv.isDarwin && !xenSupport && !nixosTestRunner && !minimal, gtk3, gettext, vte, wrapGAppsHook3
+, vncSupport ? !nixosTestRunner && !minimal, libjpeg, libpng
+, smartcardSupport ? !nixosTestRunner && !minimal, libcacard
+, spiceSupport ? true && !nixosTestRunner && !minimal, spice, spice-protocol
+, ncursesSupport ? !nixosTestRunner && !minimal, ncurses
 , usbredirSupport ? spiceSupport, usbredir
 , xenSupport ? false, xen
 , cephSupport ? false, ceph
 , glusterfsSupport ? false, glusterfs, libuuid
 , openGLSupport ? sdlSupport, mesa, libepoxy, libdrm
-, rutabagaSupport ? openGLSupport && !toolsOnly && lib.meta.availableOn stdenv.hostPlatform rutabaga_gfx, rutabaga_gfx
+, rutabagaSupport ? openGLSupport && !minimal && lib.meta.availableOn stdenv.hostPlatform rutabaga_gfx, rutabaga_gfx
 , virglSupport ? openGLSupport, virglrenderer
-, libiscsiSupport ? !toolsOnly, libiscsi
+, libiscsiSupport ? !minimal, libiscsi
 , smbdSupport ? false, samba
-, tpmSupport ? !toolsOnly
-, uringSupport ? stdenv.isLinux, liburing
-, canokeySupport ? !toolsOnly, canokey-qemu
-, capstoneSupport ? !toolsOnly, capstone
+, tpmSupport ? !minimal
+, uringSupport ? stdenv.isLinux && !userOnly, liburing
+, canokeySupport ? !minimal, canokey-qemu
+, capstoneSupport ? !minimal, capstone
 , pluginsSupport ? !stdenv.hostPlatform.isStatic
-, enableDocs ? true
-, enableTools ? true
-, enableBlobs ? true
+, enableDocs ? !minimal || toolsOnly
+, enableTools ? !minimal || toolsOnly
+, enableBlobs ? !minimal || toolsOnly
 , hostCpuOnly ? false
 , hostCpuTargets ? (if toolsOnly
                     then [ ]
                     else if hostCpuOnly
                     then (lib.optional stdenv.isx86_64 "i386-softmmu"
                           ++ ["${stdenv.hostPlatform.qemuArch}-softmmu"])
-                    else null)
+                    else if userOnly
+                    then map (system: (lib.systems.elaborate { inherit system; }).qemuArch + "-linux-user") [
+                      "aarch64-linux" "armv7l-linux"
+                      "i386-linux" "x86_64-linux"
+                      "powerpc-linux" "powerpc64-linux" "powerpc64le-linux"
+                      "riscv32-linux" "riscv64-linux"
+                    ] else null)
```

Comment:
Since this is a static list anyway, why elaborate (which IIRC is expensive), rather than listing the QEMU names directly?  They're not very likely to change.

Reply:
> which IIRC is expensive

TIL. I was just imitating the [binfmt module here](https://github.com/NixOS/nixpkgs/blob/ef6a56e11f41625d0259a86db2da07f66119f192/nixos/modules/system/boot/binfmt.nix#L32).

I poked around a bit, and from what I found, I think it's nicer to not set a target list at all (static or lib.platforms.linux) and let qemu build all `*-user` targets.

This has three advantages
* It lessens the "options are complex" problem
* If Qemu does add a target, that'll automatically be included 
* If bsd ever builds, that'll have the correct target list
But it also has two wrinkles
* The build output gets a bit large, 121 MB instead of 37
* Hexagon needs excluding, for reasons I don't fully understand

Reply:
I agree that this is nicer.  I'll have a quick look at Hexagon.

Reply:
Hexagon builds fine for me on aarch64-linux?

Reply:
Sorry, I should have explained this better. I [already default-disabled](https://github.com/NixOS/nixpkgs/pull/314998/commits/70ada6be384bc1ae4fa135d708936f32ac98a408#diff-f885f3c7293fd60915aa7c2d28eea89dea548af3d732700fa1224b0cafa55a0aR202-R203) hexagon on static tools only. If I don't do that, the build fails for both `.#legacyPackages.{x86_64,aarch64}-linux.pkgsStatic.qemu-user` with the following log:

```
Program flex found: YES (/nix/store/z1fkfc5ca0rra7m5zrh8i92mx18w9i00-flex-2.6.4/bin/flex)
Program bison found: YES 3.8.2 3.8.2 (/nix/store/abm55q8k9lsbz7kv2v68ns7hkdks2v2q-bison-3.8.2/bin/bison)
Found pkg-config: YES (/nix/store/zdvrzlvzbn9ymb0z8na50w995j8np16z-pkg-config-wrapper-0.29.2/bin/pkg-config) 0.29.2
Did not find CMake 'cmake'
Found CMake: NO
Build-time dependency glib-2.0 found: NO

../target/hexagon/meson.build:191:15: ERROR: Dependency lookup for glib-2.0 with method 'pkgconfig' failed: Could not generate libs for glib-2.0:
Package libpcre2-8 was not found in the pkg-config search path.
Perhaps you should add the directory containing `libpcre2-8.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libpcre2-8', required by 'glib-2.0', not found
```

There is a `dependency('glib-2.0', native: true)` in there. Which I thought should work since it's `native`. (rel: [`native` meson docs](https://github.com/mesonbuild/meson/blob/fa4f2339465ce3d755e2df802ebd5aa962e2ad27/docs/yaml/functions/dependency.yaml#L148-L151)). Hm.

Reply:
oh, this is only a problem when building it statically? I think that's what I was doing wrong.

Reply:
Adding pcre2 to nativeBuildInputs fixes it for me, but I think it's a bug that this is required.  Looking some more…

Reply:
I've sent this [patch](https://lore.kernel.org/qemu-devel/20240805104921.4035256-1-hi@alyssa.is/) upstream that would avoid the need to add pcre2 to nativeBuildInputs.

Reply:
> Adding pcre2 to nativeBuildInputs fixes it for me, but I think it's a bug that this is required. Looking some more…

Uff, I thought I had tried that. I think that's a fine workaround for this PR.

Thank you very much for looking this deep.